Importing some packages

Reference:

Seminar 01 (week 02): page 13

In [ ]:
import pandas as pd
import plotly.express as px

## Data Understanding & Preprocessing

Reading the original dataset

Reference:

Seminar 01 (week 02): page 13

In [ ]:
data = pd.read_csv('/content/drive/MyDrive/IIT /Level 5/Semester 2/Machine Learning & Data Mining/CW/5DATA002W.2 Coursework Dataset(25012025v6.0).csv')
data.head()

Patient_ID  Month_of_Birth   Age     Sex   Occupation T_Stage N_Stage  \
0      A0012              12  68.0  Female     Teaching      T1      N1   
1      A0013              12  50.0  Female      Medical      T2      N2   
2      A0014              11  58.0  Female  Engineering      T3      N3   
3      A0015               3  58.0  Female   Technology      T1      N1   
4      A0016               1  47.0  Female   Multimedia      T2      N1   

  6th_Stage             Differentiated  Grade   A_Stage  Tumor_Size  \
0       IIA      Poorly differentiated      3  Regional         4.0   
1      IIIA  Moderately differentiated      2  Regional        35.0   
2      IIIC  Moderately differentiated      2  Regional        63.0   
3       IIA      Poorly differentiated      3  Regional        18.0   
4       IIB      Poorly differentiated      3  Regional        41.0   

  Estrogen_Status Progesterone_Status  Regional_Node_Examined  \
0        Positive            Positive                    24.0   
1        Positive            Positive                    14.0   
2        Positive            Positive                    14.0   
3        Positive            Positive                     2.0   
4        Positive            Positive                     3.0   

   Reginol_Node_Positive  Survival_Months Mortality_Status  
0                      1               60            Alive  
1                      5               62            Alive  
2                      7               75            Alive  
3                      1               84            Alive  
4                      1               50            Alive

Printing all the columns on the data frame

Reference:

Seminar 01 (week 02): page 14

In [ ]:
list(data.columns)

['Patient_ID',
 'Month_of_Birth',
 'Age',
 'Sex',
 'Occupation',
 'T_Stage',
 'N_Stage',
 '6th_Stage',
 'Differentiated',
 'Grade',
 'A_Stage',
 'Tumor_Size',
 'Estrogen_Status',
 'Progesterone_Status',
 'Regional_Node_Examined',
 'Reginol_Node_Positive',
 'Survival_Months',
 'Mortality_Status']

Printing the summary of the data frame with data types

Reference:

Seminar 01 (week 02): page 14

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Patient_ID              4024 non-null   object 
 1   Month_of_Birth          4024 non-null   int64  
 2   Age                     4015 non-null   float64
 3   Sex                     4020 non-null   object 
 4   Occupation              43 non-null     object 
 5   T_Stage                 4024 non-null   object 
 6   N_Stage                 4024 non-null   object 
 7   6th_Stage               4024 non-null   object 
 8   Differentiated          4024 non-null   object 
 9   Grade                   4024 non-null   int64  
 10  A_Stage                 4024 non-null   object 
 11  Tumor_Size              4021 non-null   float64
 12  Estrogen_Status         4024 non-null   object 
 13  Progesterone_Status     4024 non-null   object 
 14  Regional_Node_Examined  4023 non-null   

Important decisions from the info:
- Drop Occupation cause more than 95% is missing, It is unlikely to have a strong correlation with mortality status or survival period.
- Have Impute missing values in Age(using mean or median) and Sex(using mode), cause only a small amount is missing
- Drop the rows with missing values in Tumor_Size and Regional_Node_Examined, cause these are critical medical data and tampering with them could result in bias results
- Drop Patient_ID and Month_of_Birth (Age is also available), cause they add no real value to the model
- Have to find proper encoding strategy for T_Stage, N_Stage, Grade, Estrogen_Status, Progesterone_Status, 6th_Stage, Differentiated, A_Stage
- Should scale Age, Tumor_Size, Regional_Node_Examined, Regional_Node_Positive, Survival_Months(Made the decision from the head values)

Dropping the feilds as planned

Reference:

Seminar 01 (week 02): page 16

In [ ]:
# Dropping the columns as planned
data = data.drop(columns=(['Occupation', 'Patient_ID', 'Month_of_Birth']))
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4015 non-null   float64
 1   Sex                     4020 non-null   object 
 2   T_Stage                 4024 non-null   object 
 3   N_Stage                 4024 non-null   object 
 4   6th_Stage               4024 non-null   object 
 5   Differentiated          4024 non-null   object 
 6   Grade                   4024 non-null   int64  
 7   A_Stage                 4024 non-null   object 
 8   Tumor_Size              4021 non-null   float64
 9   Estrogen_Status         4024 non-null   object 
 10  Progesterone_Status     4024 non-null   object 
 11  Regional_Node_Examined  4023 non-null   float64
 12  Reginol_Node_Positive   4024 non-null   int64  
 13  Survival_Months         4024 non-null   int64  
 14  Mortality_Status        4024 non-null   

Getting a descriptive statistics after dropping the unwaanted features

Reference:

Seminar 01 (week 02): page 14

In [ ]:
data.describe().transpose()

count       mean        std   min   25%   50%   75%  \
Age                     4015.0  54.107098  11.715528 -50.0  47.0  54.0  61.0   
Grade                   4024.0   2.150596   0.638234   1.0   2.0   2.0   3.0   
Tumor_Size              4021.0  30.419299  21.161080 -75.0  16.0  25.0  38.0   
Regional_Node_Examined  4023.0  14.373602   8.129293   1.0   9.0  14.0  19.0   
Reginol_Node_Positive   4024.0   4.158052   5.109331   1.0   1.0   2.0   5.0   
Survival_Months         4024.0  71.472167  25.361855   1.0  56.0  73.0  90.0   

                          max  
Age                     502.0  
Grade                     4.0  
Tumor_Size              140.0  
Regional_Node_Examined   61.0  
Reginol_Node_Positive    46.0  
Survival_Months         760.0

Looking for value errors in object types

Reference:

Seminar 01 (week 02): page 14

In [ ]:
data.describe(include='object').transpose()

count unique                        top  freq
Sex                  4020      2                     Female  4001
T_Stage              4024      4                         T2  1786
N_Stage              4024      3                         N1  2732
6th_Stage            4024      5                        IIA  1305
Differentiated       4024      4  Moderately differentiated  2351
A_Stage              4024      2                   Regional  3932
Estrogen_Status      4024      2                   Positive  3755
Progesterone_Status  4024      2                   Positive  3326
Mortality_Status     4024      7                      Alive  3399

Checking for missing values

Reference:

Seminar 01 (week 02): page 15

In [ ]:
data.isnull().sum()

Age                       9
Sex                       4
T_Stage                   0
N_Stage                   0
6th_Stage                 0
Differentiated            0
Grade                     0
A_Stage                   0
Tumor_Size                3
Estrogen_Status           0
Progesterone_Status       0
Regional_Node_Examined    1
Reginol_Node_Positive     0
Survival_Months           0
Mortality_Status          0
dtype: int64

Visulising target variablle for classification task

Reference:

Seminar 01 (week 02): page 24

In [ ]:
mortality_status_fig = px.bar(data, x='Mortality_Status')
mortality_status_fig.show()

## Missing Value Imputation along with Outlier Detection

Before dealing with the missing values, it is better to deal with the outliers cause outliers can effect statistical imputations

Reference:

Seminar 01 (week 02): page 23

In [ ]:
# Visualizing data to find outliers in Age
age_fig = px.histogram(data, x='Age')
age_fig.show()

There are age values like -50, 180 and 500. These are extreme outliers which can never be possible practically, So removing those outliers

Reference:

Seminar 01 (week 02): page 31

In [ ]:
# Creating a mask for age outliers (negative or > 100) that doesn't include nulls
# Getting the outliers that are not null
outlier_mask = ((data['Age'] <= 0) | (data['Age'] > 100)) & pd.notna(data['Age'])

# Counting the actual outliers (excluding nulls)
actual_outliers = outlier_mask.sum()
print(f"Actual outlier ages (excluding nulls): {actual_outliers}")

# Applying the filter
data = data[~outlier_mask]

data.info()

Actual outlier ages (excluding nulls): 3
<class 'pandas.core.frame.DataFrame'>
Index: 4021 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4012 non-null   float64
 1   Sex                     4017 non-null   object 
 2   T_Stage                 4021 non-null   object 
 3   N_Stage                 4021 non-null   object 
 4   6th_Stage               4021 non-null   object 
 5   Differentiated          4021 non-null   object 
 6   Grade                   4021 non-null   int64  
 7   A_Stage                 4021 non-null   object 
 8   Tumor_Size              4018 non-null   float64
 9   Estrogen_Status         4021 non-null   object 
 10  Progesterone_Status     4021 non-null   object 
 11  Regional_Node_Examined  4020 non-null   float64
 12  Reginol_Node_Positive   4021 non-null   int64  
 13  Survival_Months         4021 non-null   int64  
 14  Mort

Visualizing age again to make sure that the outliers have been removed

Reference:

Seminar 01 (week 02): page 23

In [ ]:
# Visualizing data to make sure that the outliers have been removed properlly
age_fig = px.histogram(data, x='Age')
age_fig.show()

In this histogram 89 might look like an outlier it is indeed a statistical outlier, but in medical setting being diagnosed with breast cancer at the age of 89 might be possible.

Reference:

BCRF. (2025). Breast Cancer in the Elderly | BCRF. Breast Cancer Research Foundation. Available from https://www.bcrf.org/about-breast-cancer/breast-cancer-elderly/.

Imputing the missing values of Age use median because, median is less sensitive to any remaining outliers cause it is the 50th percentile

Reference:

Seminar 01 (week 02): page 34

In [ ]:
# Now imputing missing values in the age
age_median = data['Age'].median()

# Impute missing values
data['Age'] = data['Age'].fillna(age_median)
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4021 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4021 non-null   float64
 1   Sex                     4017 non-null   object 
 2   T_Stage                 4021 non-null   object 
 3   N_Stage                 4021 non-null   object 
 4   6th_Stage               4021 non-null   object 
 5   Differentiated          4021 non-null   object 
 6   Grade                   4021 non-null   int64  
 7   A_Stage                 4021 non-null   object 
 8   Tumor_Size              4018 non-null   float64
 9   Estrogen_Status         4021 non-null   object 
 10  Progesterone_Status     4021 non-null   object 
 11  Regional_Node_Examined  4020 non-null   float64
 12  Reginol_Node_Positive   4021 non-null   int64  
 13  Survival_Months         4021 non-null   int64  
 14  Mortality_Status        4021 non-null   objec

Visualizing Sex in a bar plot

Reference:

Seminar 01 (week 02): page 24

In [ ]:
sex_fig = px.bar(data, x='Sex')
sex_fig.show()

From the bar plot we can clearly see that the Female are the most common type for breast cancer and the Male is mentioned as 1. fixing these issues

Reference:

Seminar 01 (week 02): page 34

In [ ]:
# Imputing missing values with mode / 'Female'
data['Sex'] = data['Sex'].fillna('Female')

data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4021 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4021 non-null   float64
 1   Sex                     4021 non-null   object 
 2   T_Stage                 4021 non-null   object 
 3   N_Stage                 4021 non-null   object 
 4   6th_Stage               4021 non-null   object 
 5   Differentiated          4021 non-null   object 
 6   Grade                   4021 non-null   int64  
 7   A_Stage                 4021 non-null   object 
 8   Tumor_Size              4018 non-null   float64
 9   Estrogen_Status         4021 non-null   object 
 10  Progesterone_Status     4021 non-null   object 
 11  Regional_Node_Examined  4020 non-null   float64
 12  Reginol_Node_Positive   4021 non-null   int64  
 13  Survival_Months         4021 non-null   int64  
 14  Mortality_Status        4021 non-null   objec

Encoding the labels of Sex variable, to improve consistencies

Reference:

Seminar 01 (week 02): page 17

In [ ]:
# encoding the Sex values
data['Sex'] = data['Sex'].map({'Female': 0 , '1': 1})

# Getting the unique values in Sex to make sure that the encoding was sucessful
data['Sex'].value_counts()

Sex
0    4002
1      19
Name: count, dtype: int64

Plotting tumor size to check for outliers

Reference:

Seminar 01 (week 02): page 23

In [ ]:
tumor_fig = px.histogram(data, x='Tumor_Size')
tumor_fig.show()

Negative tumor values are not possible at all, so removing that outlier.

And according to the official SEER documentation, the acceptable range for tumor size is 0-150 mm, The biggest tumor size is also a valid tumor size in the dataset.

Reference:

Tumor Size Over Time Recode - SEER Recodes. (2018). SEER. Available from https://seer.cancer.gov/tumorsize/.

In [ ]:
# Remove negative tumor sizes that are not negative
tumor_outlier_mask = (data['Tumor_Size'] < 0) & pd.notna(data['Tumor_Size'])
data = data[~tumor_outlier_mask]

data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4020 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4020 non-null   float64
 1   Sex                     4020 non-null   int64  
 2   T_Stage                 4020 non-null   object 
 3   N_Stage                 4020 non-null   object 
 4   6th_Stage               4020 non-null   object 
 5   Differentiated          4020 non-null   object 
 6   Grade                   4020 non-null   int64  
 7   A_Stage                 4020 non-null   object 
 8   Tumor_Size              4017 non-null   float64
 9   Estrogen_Status         4020 non-null   object 
 10  Progesterone_Status     4020 non-null   object 
 11  Regional_Node_Examined  4019 non-null   float64
 12  Reginol_Node_Positive   4020 non-null   int64  
 13  Survival_Months         4020 non-null   int64  
 14  Mortality_Status        4020 non-null   objec

Plotting tumor Size to make sure that the outliers were removed

Reference:

Seminar 01 (week 02): page 23

In [ ]:
tumor_fig = px.histogram(data, x='Tumor_Size')
tumor_fig.show()

Dropping the records with null tumors

Reference:

Seminar 01 (week 02): page 33

In [ ]:
# Tumor_Size can effect the dataset so dropping the records with null tumours
data = data.dropna(subset=['Tumor_Size'])
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4017 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4017 non-null   float64
 1   Sex                     4017 non-null   int64  
 2   T_Stage                 4017 non-null   object 
 3   N_Stage                 4017 non-null   object 
 4   6th_Stage               4017 non-null   object 
 5   Differentiated          4017 non-null   object 
 6   Grade                   4017 non-null   int64  
 7   A_Stage                 4017 non-null   object 
 8   Tumor_Size              4017 non-null   float64
 9   Estrogen_Status         4017 non-null   object 
 10  Progesterone_Status     4017 non-null   object 
 11  Regional_Node_Examined  4016 non-null   float64
 12  Reginol_Node_Positive   4017 non-null   int64  
 13  Survival_Months         4017 non-null   int64  
 14  Mortality_Status        4017 non-null   objec

Plotting regional node examined

Reference:

Seminar 01 (week 02): page 23

In [ ]:
# Visualizing regional node examined
Node_fig = px.histogram(data, x='Regional_Node_Examined')
Node_fig.show()

The Regional_Node_Examined seems to be valid with most of the patients having around 1-25 lymph nodes. Even tough there are patients with 60 nodes this can be rare medical occurances. So only removing the null values

Reference:

Seminar 01 (week 02): page 33

In [ ]:
# Regional Node Examined can effect the dataset so dropping the records with null values
data = data.dropna(subset=['Regional_Node_Examined'])
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4016 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4016 non-null   float64
 1   Sex                     4016 non-null   int64  
 2   T_Stage                 4016 non-null   object 
 3   N_Stage                 4016 non-null   object 
 4   6th_Stage               4016 non-null   object 
 5   Differentiated          4016 non-null   object 
 6   Grade                   4016 non-null   int64  
 7   A_Stage                 4016 non-null   object 
 8   Tumor_Size              4016 non-null   float64
 9   Estrogen_Status         4016 non-null   object 
 10  Progesterone_Status     4016 non-null   object 
 11  Regional_Node_Examined  4016 non-null   float64
 12  Reginol_Node_Positive   4016 non-null   int64  
 13  Survival_Months         4016 non-null   int64  
 14  Mortality_Status        4016 non-null   objec

Checking for missing values again to make sure we are on the right direction

Reference:

Seminar 01 (week 02): page 15

In [ ]:
# making sure there is no missing values at this point
data.isna().sum()/len(data)*100

Age                       0.0
Sex                       0.0
T_Stage                   0.0
N_Stage                   0.0
6th_Stage                 0.0
Differentiated            0.0
Grade                     0.0
A_Stage                   0.0
Tumor_Size                0.0
Estrogen_Status           0.0
Progesterone_Status       0.0
Regional_Node_Examined    0.0
Reginol_Node_Positive     0.0
Survival_Months           0.0
Mortality_Status          0.0
dtype: float64

## Dealing with Inconsistant Data along with Outlier Detection

Getting the usnique values from the categorical variables

Reference:

pandas.DataFrame.value_counts — pandas 1.4.4 documentation. (no date). pandas.pydata.org. Available from https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.value_counts.html.

In [ ]:
# For each categorical column checking for unique values
for col in ['T_Stage', 'N_Stage', '6th_Stage', 'Differentiated', 'A_Stage', 'Estrogen_Status', 'Progesterone_Status', 'Mortality_Status']:
    print(f"\nUnique values in {col}:")
    print(data[col].value_counts())


Unique values in T_Stage:
T_Stage
T2    1785
T1    1600
T3     529
T4     102
Name: count, dtype: int64

Unique values in N_Stage:
N_Stage
N1    2728
N2     818
N3     470
Name: count, dtype: int64

Unique values in 6th_Stage:
6th_Stage
IIA     1304
IIB     1130
IIIA    1045
IIIC     470
IIIB      67
Name: count, dtype: int64

Unique values in Differentiated:
Differentiated
Moderately differentiated    2346
Poorly differentiated        1111
Well differentiated           540
Undifferentiated               19
Name: count, dtype: int64

Unique values in A_Stage:
A_Stage
Regional    3925
Distant       91
Name: count, dtype: int64

Unique values in Estrogen_Status:
Estrogen_Status
Positive    3747
Negative     269
Name: count, dtype: int64

Unique values in Progesterone_Status:
Progesterone_Status
Positive    3319
Negative     697
Name: count, dtype: int64

Unique values in Mortality_Status:
Mortality_Status
Alive    3392
Dead      597
DEAD       10
dead        8
ALIVE       5
alive       

Based on the outpur

Reference:

GeeksforGeeks. (2020). String manipulations in Pandas DataFrame. GeeksforGeeks. Available from https://www.geeksforgeeks.org/string-manipulations-in-pandas-dataframe/.

In [ ]:
# Dealing with inconsistancies in Mortality_Status
# making everything tolower case
data['Mortality_Status'] = data['Mortality_Status'].str.lower()

# Then capitalizing the first letter
data['Mortality_Status'] = data['Mortality_Status'].str.capitalize()

# Checking again
data['Mortality_Status'].value_counts()

Mortality_Status
Alive    3401
Dead      615
Name: count, dtype: int64

Visualizing target variablle for calssification task

Reference:

Seminar 01 (week 02): page 24

In [ ]:
mortality_status_fig = px.bar(data, x='Mortality_Status')
mortality_status_fig.show()

Checking for outliers in remaining numerical columns by visualizing the variables

Reference:

Seminar 01 (week 02): page 29

In [ ]:
# Checking for outliers in remaining numerical columns
for col in ['Reginol_Node_Positive', 'Grade', 'Survival_Months']:
    # Create histogram with plotly.express
    fig = px.histogram(data, x=col)

    # Display the plot
    fig.show()

    # Basic statistics
    print(f"\nStats for {col}:")
    print(f"Min: {data[col].min()}")
    print(f"Max: {data[col].max()}")
    print(f"Mean: {data[col].mean():.2f}")
    print(f"Median: {data[col].median()}")
    print(f"Missing values: {data[col].isnull().sum()}")


Stats for Reginol_Node_Positive:
Min: 1
Max: 46
Mean: 4.15
Median: 2.0
Missing values: 0



Stats for Grade:
Min: 1
Max: 4
Mean: 2.15
Median: 2.0
Missing values: 0



Stats for Survival_Months:
Min: 1
Max: 760
Mean: 71.47
Median: 73.0
Missing values: 0


Based on the graphs for the remaining numerical columns, it is evident that Regional_Node_Possitive and Grade are good to go and are medically valid too.

Survival_Months have an error where the highest value is 760 months, which is an obvious error, so removing the record from the dataset.

In [ ]:
# Remove negative tumor sizes that are not negative
survival_months_outlier = (data['Survival_Months'] > 500)
data = data[~survival_months_outlier]

data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4015 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4015 non-null   float64
 1   Sex                     4015 non-null   int64  
 2   T_Stage                 4015 non-null   object 
 3   N_Stage                 4015 non-null   object 
 4   6th_Stage               4015 non-null   object 
 5   Differentiated          4015 non-null   object 
 6   Grade                   4015 non-null   int64  
 7   A_Stage                 4015 non-null   object 
 8   Tumor_Size              4015 non-null   float64
 9   Estrogen_Status         4015 non-null   object 
 10  Progesterone_Status     4015 non-null   object 
 11  Regional_Node_Examined  4015 non-null   float64
 12  Reginol_Node_Positive   4015 non-null   int64  
 13  Survival_Months         4015 non-null   int64  
 14  Mortality_Status        4015 non-null   objec

Visualizing survival months after removing outliers to make sure the change was done

Reference:

Seminar 01 (week 02): page 29

In [ ]:
# Visualizing survival months
survival_months_fig = px.histogram(data, x='Survival_Months')
survival_months_fig.show()

### Encoding object data types

Renaming Estrogen Status and Progesterone Status to numeric type

Reference:

Seminar 01 (week 02): page 17

In [ ]:
# Binary encoding for hormone status
data['Estrogen_Status'] = data['Estrogen_Status'].map({'Positive': 1, 'Negative': 0})
data['Progesterone_Status'] = data['Progesterone_Status'].map({'Positive': 1, 'Negative': 0})
print(data['Progesterone_Status'].unique(), " : Progesterone_Status")
print(data['Progesterone_Status'].unique(), " : Progesterone_Status")

[1 0]  : Progesterone_Status
[1 0]  : Progesterone_Status


Renaming Mortality Status to numeric type

Reference:

Seminar 01 (week 02): page 17

In [ ]:
data['Mortality_Status'] = data['Mortality_Status'].map( { 'Alive': 1, 'Dead': 0 } )
print(data['Mortality_Status'].unique(), " : Mortality_Status")

[1 0]  : Mortality_Status


Renaming T Stage and N Stage to  numeric type

Reference:

Seminar 01 (week 02): page 17

In [ ]:
data['T_Stage'] = data['T_Stage'].map({'T1': 1, 'T2': 2, 'T3': 3, 'T4': 4})
data['N_Stage'] = data['N_Stage'].map({'N1': 1, 'N2': 2, 'N3': 3})
print(data['T_Stage'].unique(), " : T_Stage")
print(data['N_Stage'].unique(), " : N_Stage")

[1 2 3 4]  : T_Stage
[1 2 3]  : N_Stage


Renaming A Stage to numeric type

Reference:

Seminar 01 (week 02): page 17

In [ ]:
# Simple mapping approach
data['A_Stage'] = data['A_Stage'].map({'Regional': 0, 'Distant': 1})
data['A_Stage'].unique()

array([0, 1])

Renaming 6th Stage to numeric type

Reference:

Seminar 01 (week 02): page 17

In [ ]:
# Define a severity-based mapping for 6th_Stage
stage_severity = {
    'IIA': 0,  # Least severe in the dataset
    'IIB': 1,
    'IIIA': 2,
    'IIIB': 3,
    'IIIC': 4   # Most severe in the dataset
}

# Apply the mapping
data['6th_Stage'] = data['6th_Stage'].map(stage_severity)
data['6th_Stage'].unique()

array([0, 2, 4, 1, 3])

Renaming Differentiated to numeric type

Reference:

Seminar 01 (week 02): page 17

In [ ]:
# Define a differentiation-based mapping (from most to least differentiated)
differentiation_scale = {
    'Well differentiated': 0,      # Best prognosis
    'Moderately differentiated': 1,
    'Poorly differentiated': 2,
    'Undifferentiated': 3          # Worst prognosis
}

# Apply the mapping
data['Differentiated'] = data['Differentiated'].map(differentiation_scale)
data['Differentiated'].unique()

array([2, 1, 0, 3])

Printing the summary of the data frame with data types

Reference:

Seminar 01 (week 02): page 14

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4015 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4015 non-null   float64
 1   Sex                     4015 non-null   int64  
 2   T_Stage                 4015 non-null   int64  
 3   N_Stage                 4015 non-null   int64  
 4   6th_Stage               4015 non-null   int64  
 5   Differentiated          4015 non-null   int64  
 6   Grade                   4015 non-null   int64  
 7   A_Stage                 4015 non-null   int64  
 8   Tumor_Size              4015 non-null   float64
 9   Estrogen_Status         4015 non-null   int64  
 10  Progesterone_Status     4015 non-null   int64  
 11  Regional_Node_Examined  4015 non-null   float64
 12  Reginol_Node_Positive   4015 non-null   int64  
 13  Survival_Months         4015 non-null   int64  
 14  Mortality_Status        4015 non-null   int64

Things to consider now:
- All of the features have been encoded now and are ready for scaling
- Have to do encoding scaling on the numerical values with proper scalling mechanism
- Have to do scalling on ordinal features which have non-binary encoding

## Data Scaling

Have to scale the categorical feature variables with non-binary encoding (T_Stage, N_Stage, 6th_Stage, Differentiated) and ordinal variable (Grade)

Reason: Non-scaled variables can result in biased decision makings in KNN (Distance-based, needs scaled features to avoid bias), LR and NB (Assumes normal distribution, scale to meet that assumption)

StandardScaler to scale these variables because,
- Helps distance-based models (like KNN) treat all features fairly
- Helps Logistic Regression converge faster and make better predictions
In summary it is the best option for distance based models to have a non-bias prediction

Binary encoded categorical features don't require scaling cause they have no impact on Distance Metrics, and the distribution is not very huge comparing to the other features after scaling

Reference:

Seminar 01 (week 02): page 35

In [ ]:
from sklearn.preprocessing import StandardScaler

# Create a single scaler instance
scaler = StandardScaler()

# Define the categorical features to scale
categorical_features = ['T_Stage', 'N_Stage', '6th_Stage', 'Differentiated', 'Grade']

# Apply scaling to all these features at once
data[categorical_features] = scaler.fit_transform(data[categorical_features])

Doing Standard Scaling on the remaining numeric/continuous features too for the above mentioned same reasons

Reference:

Seminar 01 (week 02): page 35

In [ ]:
# Define remaining numeric features
numeric_features = ['Age', 'Tumor_Size', 'Regional_Node_Examined', 'Reginol_Node_Positive']

# Create a single scaler for numeric features
numeric_scaler = StandardScaler()

# Apply scaling to all numeric features at once
data[numeric_features] = numeric_scaler.fit_transform(data[numeric_features])

Getting the descriptive statistics to make sure that the data is scaled properlly

Reference:

Seminar 01 (week 02): page 28

In [ ]:
# Making sure that the data is scaled properlly
data.describe().transpose()

count          mean        std       min        25%  \
Age                     4015.0 -8.406172e-17   1.000125 -2.677136  -0.780989   
Sex                     4015.0  4.732254e-03   0.068637  0.000000   0.000000   
T_Stage                 4015.0 -1.008741e-16   1.000125 -1.024893  -1.024893   
N_Stage                 4015.0 -1.221107e-16   1.000125 -0.631472  -0.631472   
6th_Stage               4015.0  7.432826e-17   1.000125 -1.042623  -1.042623   
Differentiated          4015.0 -8.760116e-17   1.000125 -1.804953  -0.237720   
Grade                   4015.0 -8.760116e-17   1.000125 -1.804953  -0.237720   
A_Stage                 4015.0  2.266501e-02   0.148852  0.000000   0.000000   
Tumor_Size              4015.0  7.963742e-18   1.000125 -1.397007  -0.685159   
Estrogen_Status         4015.0  9.330012e-01   0.250051  0.000000   1.000000   
Progesterone_Status     4015.0  8.264010e-01   0.378811  0.000000   1.000000   
Regional_Node_Examined  4015.0 -6.813423e-17   1.000125 -1.644022  -0.659662   
Reginol_Node_Positive   4015.0 -7.963742e-17   1.000125 -0.618154  -0.618154   
Survival_Months         4015.0  7.130212e+01  22.925447  1.000000  56.000000   
Mortality_Status        4015.0  8.468244e-01   0.360202  0.000000   1.000000   

                              50%        75%         max  
Age                     -0.000222   0.780544    3.903610  
Sex                      0.000000   0.000000    1.000000  
T_Stage                  0.282269   0.282269    2.896591  
N_Stage                 -0.631472   0.812355    2.256182  
6th_Stage               -0.252339   0.537944    2.118510  
Differentiated          -0.237720   1.329513    2.896746  
Grade                   -0.237720   1.329513    2.896746  
A_Stage                  0.000000   0.000000    1.000000  
Tumor_Size              -0.258050   0.358885    5.199452  
Estrogen_Status          1.000000   1.000000    1.000000  
Progesterone_Status      1.000000   1.000000    1.000000  
Regional_Node_Examined  -0.044437   0.570788    5.738677  
Reginol_Node_Positive   -0.421849   0.167067    8.215584  
Survival_Months         73.000000  90.000000  107.000000  
Mortality_Status         1.000000   1.000000    1.000000

## Exporting Data

Due to data leakage it is better to remove Survival_Months for the classification dataset and Mortality_Status for the Regression dataset, as they would directly reveal the outcome the model is trying to predict. Doing so will prevent data leakage.

Same concept has been discussed on Tutorial week 06

Reference:

 - GeeksforGeeks. (2024). What is Data Leakage? GeeksforGeeks. Available from https://www.geeksforgeeks.org/what-is-data-leakage/.
 - Seminar 06 (week 07)

In [ ]:
# Create classification dataset by removing Survival_Months
classification_data = data.drop(columns=['Survival_Months'])

classification_data.to_csv('classification_data.csv', index=False)

Getting descriptive statistics about the classification dataset after removing the survival months

Reference:

Seminar 01 (week 02): page 28

In [ ]:
classification_data.describe().transpose()

count          mean       std       min       25%  \
Age                     4015.0 -8.406172e-17  1.000125 -2.677136 -0.780989   
Sex                     4015.0  4.732254e-03  0.068637  0.000000  0.000000   
T_Stage                 4015.0 -1.008741e-16  1.000125 -1.024893 -1.024893   
N_Stage                 4015.0 -1.221107e-16  1.000125 -0.631472 -0.631472   
6th_Stage               4015.0  7.432826e-17  1.000125 -1.042623 -1.042623   
Differentiated          4015.0 -8.760116e-17  1.000125 -1.804953 -0.237720   
Grade                   4015.0 -8.760116e-17  1.000125 -1.804953 -0.237720   
A_Stage                 4015.0  2.266501e-02  0.148852  0.000000  0.000000   
Tumor_Size              4015.0  7.963742e-18  1.000125 -1.397007 -0.685159   
Estrogen_Status         4015.0  9.330012e-01  0.250051  0.000000  1.000000   
Progesterone_Status     4015.0  8.264010e-01  0.378811  0.000000  1.000000   
Regional_Node_Examined  4015.0 -6.813423e-17  1.000125 -1.644022 -0.659662   
Reginol_Node_Positive   4015.0 -7.963742e-17  1.000125 -0.618154 -0.618154   
Mortality_Status        4015.0  8.468244e-01  0.360202  0.000000  1.000000   

                             50%       75%       max  
Age                    -0.000222  0.780544  3.903610  
Sex                     0.000000  0.000000  1.000000  
T_Stage                 0.282269  0.282269  2.896591  
N_Stage                -0.631472  0.812355  2.256182  
6th_Stage              -0.252339  0.537944  2.118510  
Differentiated         -0.237720  1.329513  2.896746  
Grade                  -0.237720  1.329513  2.896746  
A_Stage                 0.000000  0.000000  1.000000  
Tumor_Size             -0.258050  0.358885  5.199452  
Estrogen_Status         1.000000  1.000000  1.000000  
Progesterone_Status     1.000000  1.000000  1.000000  
Regional_Node_Examined -0.044437  0.570788  5.738677  
Reginol_Node_Positive  -0.421849  0.167067  8.215584  
Mortality_Status        1.000000  1.000000  1.000000

Printing all the columns on the classification data

Reference:

Seminar 01 (week 02): page 14

In [ ]:
list(classification_data.columns)

['Age',
 'Sex',
 'T_Stage',
 'N_Stage',
 '6th_Stage',
 'Differentiated',
 'Grade',
 'A_Stage',
 'Tumor_Size',
 'Estrogen_Status',
 'Progesterone_Status',
 'Regional_Node_Examined',
 'Reginol_Node_Positive',
 'Mortality_Status']

Printing the dimensions of the classification data frame

Reference:

Seminar 01 (week 02): page 15

In [ ]:
classification_data.shape

(4015, 14)

For the regression task specifications are asking for, **For patients who would not survive cancer**, can machine learning offer a reliable estimate of their survival period?

So it is mandatory to remove the records of people who are alive, and work with records of people who are deceased.

So for the regression dataset we have to
1. remove the records with people who are alive
2. Remove the mortality_status field

Reference:

Seminar 06 (week 07): page 12

In [ ]:
# Filtering to keep only patients who didn't survive (Mortality_Status = 0)
# Creating a new dataframe called deceased_patients
deceased_patients = data[(data.Mortality_Status == 0)]

# Check the basic stats to observe only deceased patients remain
deceased_patients.describe().transpose()

count       mean        std       min        25%  \
Age                     615.0   0.130540   1.081362 -2.677136  -0.669451   
Sex                     615.0   0.001626   0.040324  0.000000   0.000000   
T_Stage                 615.0   0.365162   1.094574 -1.024893  -1.024893   
N_Stage                 615.0   0.601063   1.217830 -0.631472  -0.631472   
6th_Stage               615.0   0.606049   1.129594 -1.042623  -0.252339   
Differentiated          615.0   0.378980   0.985486 -1.804953  -0.237720   
Grade                   615.0   0.378980   0.985486 -1.804953  -0.237720   
A_Stage                 615.0   0.055285   0.228721  0.000000   0.000000   
Tumor_Size              615.0   0.318682   1.145293 -1.397007  -0.495333   
Estrogen_Status         615.0   0.824390   0.380798  0.000000   1.000000   
Progesterone_Status     615.0   0.668293   0.471210  0.000000   0.000000   
Regional_Node_Examined  615.0   0.078208   1.042772 -1.644022  -0.659662   
Reginol_Node_Positive   615.0   0.600535   1.427220 -0.618154  -0.421849   
Survival_Months         615.0  45.619512  23.984691  2.000000  27.000000   
Mortality_Status        615.0   0.000000   0.000000  0.000000   0.000000   

                              50%        75%         max  
Age                      0.334392   1.003620    1.672849  
Sex                      0.000000   0.000000    1.000000  
T_Stage                  0.282269   1.589430    2.896591  
N_Stage                  0.812355   2.256182    2.256182  
6th_Stage                0.537944   2.118510    2.118510  
Differentiated          -0.237720   1.329513    2.896746  
Grade                   -0.237720   1.329513    2.896746  
A_Stage                  0.000000   0.000000    1.000000  
Tumor_Size              -0.020767   0.928363    5.199452  
Estrogen_Status          1.000000   1.000000    1.000000  
Progesterone_Status      1.000000   1.000000    1.000000  
Regional_Node_Examined  -0.044437   0.693833    5.246497  
Reginol_Node_Positive   -0.029238   1.148594    8.215584  
Survival_Months         44.000000  61.000000  102.000000  
Mortality_Status         0.000000   0.000000    0.000000

Exporting the regression dataset after dropping Mortality Status

Reference:

Seminar 06 (week 07): page 12

In [ ]:
# Create regression dataset by removing Survival_Months
regression_data = deceased_patients.drop(columns=['Mortality_Status'])

regression_data.to_csv('regression_data.csv', index=False)

Getting descriptive statistics about the regression dataset after removing the Mortality Status

Reference:

Seminar 01 (week 02): page 28

In [ ]:
regression_data.describe().transpose()

count       mean        std       min        25%  \
Age                     615.0   0.130540   1.081362 -2.677136  -0.669451   
Sex                     615.0   0.001626   0.040324  0.000000   0.000000   
T_Stage                 615.0   0.365162   1.094574 -1.024893  -1.024893   
N_Stage                 615.0   0.601063   1.217830 -0.631472  -0.631472   
6th_Stage               615.0   0.606049   1.129594 -1.042623  -0.252339   
Differentiated          615.0   0.378980   0.985486 -1.804953  -0.237720   
Grade                   615.0   0.378980   0.985486 -1.804953  -0.237720   
A_Stage                 615.0   0.055285   0.228721  0.000000   0.000000   
Tumor_Size              615.0   0.318682   1.145293 -1.397007  -0.495333   
Estrogen_Status         615.0   0.824390   0.380798  0.000000   1.000000   
Progesterone_Status     615.0   0.668293   0.471210  0.000000   0.000000   
Regional_Node_Examined  615.0   0.078208   1.042772 -1.644022  -0.659662   
Reginol_Node_Positive   615.0   0.600535   1.427220 -0.618154  -0.421849   
Survival_Months         615.0  45.619512  23.984691  2.000000  27.000000   

                              50%        75%         max  
Age                      0.334392   1.003620    1.672849  
Sex                      0.000000   0.000000    1.000000  
T_Stage                  0.282269   1.589430    2.896591  
N_Stage                  0.812355   2.256182    2.256182  
6th_Stage                0.537944   2.118510    2.118510  
Differentiated          -0.237720   1.329513    2.896746  
Grade                   -0.237720   1.329513    2.896746  
A_Stage                  0.000000   0.000000    1.000000  
Tumor_Size              -0.020767   0.928363    5.199452  
Estrogen_Status          1.000000   1.000000    1.000000  
Progesterone_Status      1.000000   1.000000    1.000000  
Regional_Node_Examined  -0.044437   0.693833    5.246497  
Reginol_Node_Positive   -0.029238   1.148594    8.215584  
Survival_Months         44.000000  61.000000  102.000000

Printing all the columns on the regression data frame

Reference:

Seminar 01 (week 02): page 14

In [ ]:
list(regression_data.columns)

['Age',
 'Sex',
 'T_Stage',
 'N_Stage',
 '6th_Stage',
 'Differentiated',
 'Grade',
 'A_Stage',
 'Tumor_Size',
 'Estrogen_Status',
 'Progesterone_Status',
 'Regional_Node_Examined',
 'Reginol_Node_Positive',
 'Survival_Months']

Printing the dimensions of the regression data frame

Reference:

Seminar 01 (week 02): page 15

In [ ]:
regression_data.shape

(615, 14)